### ISSUE TITLE
[3.0 Feature Request]: iOS Delegate Splitter

### ISSUE THREAD

**Issue by vbresults** [NativeScript/NativeScript#4065]
There are a lot of plugins out there that explicitly set the application delegate and it is extremely troublesome getting them working together.

I propose a "splitter" to split/broadcast the delegate event asynchronously.  The class I use for all of my custom plugins looks like this (which I point `application.ios.delegate` to):
```typescript
export class Delegate extends UIResponder implements UIApplicationDelegate {
	public static ObjCProtocols = [UIApplicationDelegate];
	protected static _queue: Object = {};

	applicationOpenURLOptions(application: typeof UIApplication, url: string, options: typeof NSDictionary) {
		return Delegate._promise("applicationOpenURLOptions", { application: application, url: url, options: options });
	}

	applicationDidFinishLaunchingWithOptions(application: typeof UIApplication, launchOptions: any) {
		return Delegate._promise("applicationDidFinishLaunchingWithOptions", { application: application, launchOptions: launchOptions });
	}

	// etc...

	protected static _promise(fn: string, args: any) {
		let constants = {
			state: args.application.applicationState
		};

		let promise = Promise.resolve().then(() => {
			return { fn: fn, args: args, constants: constants };
		});

		let entry;

		if (!(entry = this._queue[fn])) {
			entry = { callbacks: [], promise: promise };
			this._queue[fn] = entry;
			return entry;
		}

		entry.promise = promise;

		if (entry.callbacks.length > 0) {
			entry.callbacks.forEach(function(callback) {
				entry.promise.then(callback);
			});
		}

		return entry;
	}

	public static apply(fn: string, callback: any) {
		let entry;

		if (!(entry = this._queue[fn])) {
			entry = this._queue[fn] = { callbacks: [], promise: false };
		}

		if (!entry.promise) {
			entry.callbacks.push(callback);
		} else {
			entry.promise.then(callback);
		}

		return entry;
	}
}
```

Then in my plugins, I use:
```typescript
Delegate.apply("applicationDidFinishLaunchingWithOptions", (event) => {
	// event.args contains applicationDidFinishLaunchingWithOptions native arguments...
});
```

Works great for me.  What are the thoughts on this concept as a core feature?

**Comment by PanayotCankov** [comment]
I am fan of the idea.
I am linking the last time there was some development on app delegates:
https://github.com/NativeScript/NativeScript/issues/432
I believe we can make plugins register protocols and methods (event like) on a common app delegate before the application.start() is called.

Here is one plugin I know merging AppDelegates:
https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/master/firebase.ios.js#L291

There are probably more, can we compile a list?

**Comment by sitefinitysteve** [comment]
I very much need "applicationOpenURLOptions" for the Auth0 plugin... @vbresults sample seems to do the callback just fine, but whats the plan going forward to make this a bit easier to implement for users?

**Comment by hypery2k** [comment]
you got give also my plugin a try for ease URL handling: https://github.com/hypery2k/nativescript-urlhandler

**Comment by Mitko-Kerezov** [comment]
The approach taken by @hypery2k seems the best one in my opinion, however currently it has some flaws because of a lack of a unified way to manage the application's delegate.

The approach revolves around trying to *chain* the methods with one another [like so](https://github.com/hypery2k/nativescript-urlhandler/blob/959f909ea5dba01d73ac240642d303fddcdc7bd1/src/urlhandler.ios.ts#L7). This idea is exceptional, however is quickly undermined by something like the [firebase](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/6d81955a01fcab057e54a51a56cd60583df34373/src/firebase.ios.ts#L48) or the [facebook](https://github.com/NativeScript/nativescript-facebook/blob/d5bf487bc75fd3477580821fa47f364860276973/src/index.ios.ts#L30) plugins.

My proposition revolves around the same idea backed up by some code in the `tns-core-modules`.

1. Adapt the [chain logic](https://github.com/hypery2k/nativescript-urlhandler/blob/959f909ea5dba01d73ac240642d303fddcdc7bd1/src/urlhandler.ios.ts#L7-L28) inside the tns-core-modules
2. Modify the `require("application").ios.delegate` [**setter**](https://github.com/NativeScript/NativeScript/blob/af651d6e831ce02402eaf097613ec67742b01d52/tns-core-modules/application/application.ios.ts#L96-L100) so that it becomes something like this:

```JS
set delegate(value: typeof UIApplicationDelegate) {
    Object.getOwnPropertyNames(value).forEach(methodName => {
        enableMultipleOverridesFor(this._delegate, methodName);
    });
}
```

The benefits of this approach would be:
1. One single instance of the application delegate with no way for external code (that is outside of tns-core-modules) to **set** or directly **modify** it directly (meaning no way for plugins to hinder each other's functionality anymore)
2. Backwards compatibility - code like `require("application").ios.delegate = MyCustomDelegate;` will start working out-of-the-box.

The drawback of the approach would be:
1. No way of controlling the order of the method invocations
2. Code altering methods [directly](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/6d81955a01fcab057e54a51a56cd60583df34373/src/firebase.ios.ts#L48) will have to be handled in some way

**Comment by Naxos84** [comment]
I don't know why. But the url-handler plugin seems not to work when also using https://github.com/alexziskind1/nativescript-oauth2

**Comment by SylannBin** [comment]
@Naxos84 this is because the current setter of IOSApplication.delegate will only accept a first value and discard following assignation attempts.
@hypery2k 's plugin (url-handler) tries to conserve existing methods of a previous delegate implementation such that all methods will be executed one after the other. but this relies on the assumption that the delegate implementation is thought as a method container whit no state.
@alexziskind1's Oauth2 plugin doesn't care about previous implementations and will just try to set the delegate with a class where methods actually need a specific state. This implementation is plainly incompatible with any other, and I was quite disappointed to see that in a plugin.

**Comment by SylannBin** [comment]
I thought again about this and I wonder if it wouldn't be better to just remove the setter guard to allow people to keep control over the delegate. Maybe plugins should only propose a set of methods that the user has the responsibility to put in the delegate. Any thought on this?

**Comment by alexziskind1** [comment]
@SylannBin yes this is a known issue. A PR to fix would be greatly appreciated. Thanks!
